diff --git a/node_modules/avsc/lib/types.js b/node_modules/avsc/lib/types.js
index dc7c69c..e399f82 100644
--- a/node_modules/avsc/lib/types.js
+++ b/node_modules/avsc/lib/types.js
@@ -97,17 +97,17 @@ function Type(schema, opts) {
       name = maybeQualify(name, namespace);
       if (isPrimitive(name)) {
         // Avro doesn't allow redefining primitive names.
-        throw new Error(f('cannot rename primitive type: %j', name));
+        Type.addError(new Error(f('cannot rename primitive type: %j', name)));
       }
       var registry = opts && opts.registry;
       if (registry) {
         if (registry[name] !== undefined) {
-          throw new Error(f('duplicate type name: %s', name));
+          Type.addError(new Error(f('duplicate type name: %s', name)));
         }
         registry[name] = type;
       }
     } else if (opts && opts.noAnonymousTypes) {
-      throw new Error(f('missing name property in schema: %j', schema));
+      Type.addError(new Error(f('missing name property in schema: %j', schema)));
     }
     this.name = name;
     this.aliases = schema.aliases ?
@@ -138,13 +138,13 @@ Type.forSchema = function (schema, opts) {
       case 'auto':
         return undefined; // Determined dynamically later on.
       default:
-        throw new Error(f('invalid wrap unions option: %j', wrapUnions));
+        Type.addError(new Error(f('invalid wrap unions option: %j', wrapUnions)));
     }
   })(opts.wrapUnions);
 
   if (schema === null) {
     // Let's be helpful for this common error.
-    throw new Error('invalid type: null (did you mean "null"?)');
+    Type.addError(new Error('invalid type: null (did you mean "null"?)'));
   }
 
   if (Type.isType(schema)) {
@@ -154,7 +154,7 @@ Type.forSchema = function (schema, opts) {
   var type;
   if (opts.typeHook && (type = opts.typeHook(schema, opts))) {
     if (!Type.isType(type)) {
-      throw new Error(f('invalid typehook return value: %j', type));
+      Type.addError(new Error(f('invalid typehook return value: %j', type)));
     }
     return type;
   }
@@ -170,9 +170,9 @@ Type.forSchema = function (schema, opts) {
       // Reference to a primitive type. These are also defined names by default
       // so we create the appropriate type and it to the registry for future
       // reference.
-      return opts.registry[schema] = Type.forSchema({type: schema}, opts);
-    }
-    throw new Error(f('undefined type name: %s', schema));
+      return (opts.registry[schema] = Type.forSchema({type: schema}, opts));
+    }// Type.addError(new Error(f('undefined type name: %s', schema))); - return schema directly
+    return schema;
   }
 
   if (schema.logicalType && opts.logicalTypes && !LOGICAL_TYPE) {
@@ -219,7 +219,7 @@ Type.forSchema = function (schema, opts) {
     type = (function (typeName) {
       var Type = TYPES[typeName];
       if (Type === undefined) {
-        throw new Error(f('unknown type: %j', typeName));
+        Type.addError(new Error(f('unknown type: %j', typeName)));
       }
       return new Type(schema, opts);
     })(schema.type);
@@ -240,7 +240,7 @@ Type.forValue = function (val, opts) {
     var type = opts.valueHook(val, opts);
     if (type !== undefined) {
       if (!Type.isType(type)) {
-        throw new Error(f('invalid value hook return value: %j', type));
+        Type.addError( new Error(f('invalid value hook return value: %j', type)));
       }
       return type;
     }
@@ -293,13 +293,13 @@ Type.forValue = function (val, opts) {
         })
       }, opts);
     default:
-      throw new Error(f('cannot infer type from: %j', val));
+      Type.addError(new Error(f('cannot infer type from: %j', val)));
   }
 };
 
 Type.forTypes = function (types, opts) {
   if (!types.length) {
-    throw new Error('no types to combine');
+    Type.addError(new Error('no types to combine'));
   }
   if (types.length === 1) {
     return types[0]; // Nothing to do.
@@ -333,7 +333,7 @@ Type.forTypes = function (types, opts) {
       // It is only valid to combine wrapped unions when no other type is
       // present other than wrapped unions and nulls (otherwise the values of
       // others wouldn't be valid in the resulting union).
-      throw new Error('cannot combine wrapped union');
+      Type.addError(new Error('cannot combine wrapped union'));
     }
     var branchTypes = {};
     expanded.forEach(function (type) {
@@ -342,7 +342,7 @@ Type.forTypes = function (types, opts) {
       if (!branchType) {
         branchTypes[name] = type;
       } else if (!type.equals(branchType)) {
-        throw new Error('inconsistent branch type');
+        Type.addError(new Error('inconsistent branch type'));
       }
     });
     var wrapUnions = opts.wrapUnions;
@@ -496,7 +496,7 @@ Type.prototype.createResolver = function (type, opts) {
   if (!Type.isType(type)) {
     // More explicit error message than the "incompatible type" thrown
     // otherwise (especially because of the overridden `toJSON` method).
-    throw new Error(f('not a type: %j', type));
+    Type.addError(new Error(f('not a type: %j', type)));
   }
 
   if (!Type.isType(this, 'union', 'logical') && Type.isType(type, 'logical')) {
@@ -599,10 +599,10 @@ Type.prototype.fromBuffer = function (buf, resolver, noCheck) {
   var tap = new Tap(buf);
   var val = readValue(this, tap, resolver, noCheck);
   if (!tap.isValid()) {
-    throw new Error('truncated buffer');
+    Type.addError(new Error('truncated buffer'));
   }
   if (!noCheck && tap.pos < buf.length) {
-    throw new Error('trailing data');
+    Type.addError(new Error('trailing data'));
   }
   return val;
 };
@@ -914,7 +914,7 @@ LongType.prototype._check = function (val, flags, hook) {
 LongType.prototype._read = function (tap) {
   var n = tap.readLong();
   if (!isSafeLong(n)) {
-    throw new Error('potential precision loss');
+    Type.addError(new Error('potential precision loss'));
   }
   return n;
 };
@@ -963,7 +963,7 @@ LongType.__with = function (methods, noUnpack) {
   var type = new AbstractLongType(noUnpack);
   Object.keys(mapping).forEach(function (name) {
     if (methods[name] === undefined) {
-      throw new Error(f('missing method implementation: %s', name));
+      Type.addError(new Error(f('missing method implementation: %s', name)));
     }
     type[mapping[name]] = methods[name];
   });
@@ -1147,14 +1147,14 @@ BytesType.prototype._copy = function (obj, opts) {
       return obj.toString('binary');
     case 2: // Coerce strings to buffers.
       if (typeof obj != 'string') {
-        throw new Error(f('cannot coerce to buffer: %j', obj));
+        Type.addError(new Error(f('cannot coerce to buffer: %j', obj)));
       }
       buf = utils.bufferFrom(obj, 'binary');
       this._check(buf, undefined, throwInvalidError);
       return buf;
     case 1: // Coerce buffer JSON representation to buffers.
       if (!isJsonBuffer(obj)) {
-        throw new Error(f('cannot coerce to buffer: %j', obj));
+        Type.addError(new Error(f('cannot coerce to buffer: %j', obj)));
       }
       buf = utils.bufferFrom(obj.data);
       this._check(buf, undefined, throwInvalidError);
@@ -1178,10 +1178,10 @@ function UnionType(schema, opts) {
   Type.call(this);
 
   if (!Array.isArray(schema)) {
-    throw new Error(f('non-array union schema: %j', schema));
+    Type.addError(new Error(f('non-array union schema: %j', schema)));
   }
   if (!schema.length) {
-    throw new Error('empty union');
+    Type.addError(new Error('empty union'));
   }
   this.types = Object.freeze(schema.map(function (obj) {
     return Type.forSchema(obj, opts);
@@ -1190,11 +1190,11 @@ function UnionType(schema, opts) {
   this._branchIndices = {};
   this.types.forEach(function (type, i) {
     if (Type.isType(type, 'union')) {
-      throw new Error('unions cannot be directly nested');
+      Type.addError(new Error('unions cannot be directly nested'));
     }
     var branch = type.branchName;
-    if (this._branchIndices[branch] !== undefined) {
-      throw new Error(f('duplicate union branch name: %j', branch));
+    if (branch && this._branchIndices[branch] !== undefined) {
+      Type.addError(new Error(f('duplicate union branch name: %j', branch)));
     }
     this._branchIndices[branch] = i;
   }, this);
@@ -1202,7 +1202,7 @@ function UnionType(schema, opts) {
 util.inherits(UnionType, Type);
 
 UnionType.prototype._branchConstructor = function () {
-  throw new Error('unions cannot be directly wrapped');
+  Type.addError(new Error('unions cannot be directly wrapped'));
 };
 
 UnionType.prototype._skip = function (tap) {
@@ -1258,7 +1258,7 @@ function UnwrappedUnionType(schema, opts) {
     } else {
       var bucket = getTypeBucket(type);
       if (this._bucketIndices[bucket] !== undefined) {
-        throw new Error(f('ambiguous unwrapped union: %j', this));
+        Type.addError(new Error(f('ambiguous unwrapped union: %j', this)));
       }
       this._bucketIndices[bucket] = index;
     }
@@ -1289,7 +1289,7 @@ UnwrappedUnionType.prototype._getBranchIndex = function (any, index) {
         // More than one branch matches the value so we aren't guaranteed to
         // infer the correct type. We throw rather than corrupt data. This can
         // be fixed by "tightening" the logical types.
-        throw new Error('ambiguous conversion');
+        Type.addError(new Error('ambiguous conversion'));
       }
     }
   }
@@ -1314,7 +1314,7 @@ UnwrappedUnionType.prototype._read = function (tap) {
   if (branchType) {
     return branchType._read(tap);
   } else {
-    throw new Error(f('invalid union index: %s', index));
+    Type.addError(new Error(f('invalid union index: %s', index)));
   }
 };
 
@@ -1379,7 +1379,7 @@ UnwrappedUnionType.prototype._copy = function (val, opts) {
         index = this._getIndex(val);
     }
     if (index === undefined) {
-      throwInvalidError(val, this);
+      throwInvalidError(val, this); return;
     }
   }
   var type = this.types[index];
@@ -1479,7 +1479,7 @@ WrappedUnionType.prototype._check = function (val, flags, hook, path) {
 WrappedUnionType.prototype._read = function (tap) {
   var type = this.types[tap.readLong()];
   if (!type) {
-    throw new Error(f('invalid union index'));
+    Type.addError(new Error(f('invalid union index')));
   }
   var Branch = type._branchConstructor;
   if (Branch === null) {
@@ -1629,22 +1629,22 @@ WrappedUnionType.prototype.random = function () {
 function EnumType(schema, opts) {
   Type.call(this, schema, opts);
   if (!Array.isArray(schema.symbols) || !schema.symbols.length) {
-    throw new Error(f('invalid enum symbols: %j', schema.symbols));
+    Type.addError(new Error(f('invalid enum symbols: %j', schema.symbols)));
   }
   this.symbols = Object.freeze(schema.symbols.slice());
   this._indices = {};
   this.symbols.forEach(function (symbol, i) {
     if (!utils.isValidName(symbol)) {
-      throw new Error(f('invalid %s symbol: %j', this, symbol));
+      Type.addError(new Error(f('invalid %s symbol: %j', this, symbol)));
     }
     if (this._indices[symbol] !== undefined) {
-      throw new Error(f('duplicate %s symbol: %j', this, symbol));
+      Type.addError(new Error(f('duplicate %s symbol: %j', this, symbol)));
     }
     this._indices[symbol] = i;
   }, this);
   this.default = schema.default;
   if (this.default !== undefined && this._indices[this.default] === undefined) {
-    throw new Error(f('invalid %s default: %j', this, this.default));
+    Type.addError(new Error(f('invalid %s default: %j', this, this.default)));
   }
   this._branchConstructor = this._createBranchConstructor();
   Object.freeze(this);
@@ -1663,7 +1663,7 @@ EnumType.prototype._read = function (tap) {
   var index = tap.readLong();
   var symbol = this.symbols[index];
   if (symbol === undefined) {
-    throw new Error(f('invalid %s enum index: %s', this.name, index));
+    Type.addError(new Error(f('invalid %s enum index: %s', this.name, index)));
   }
   return symbol;
 };
@@ -1723,8 +1723,8 @@ EnumType.prototype.random = function () {
 /** Avro fixed type. Represented simply as a `Buffer`. */
 function FixedType(schema, opts) {
   Type.call(this, schema, opts);
-  if (schema.size !== (schema.size | 0) || schema.size < 0) {
-    throw new Error(f('invalid %s size', this.branchName));
+  if (schema.size !== (schema.size | 0) || schema.size < 1) {
+    Type.addError(new Error(f('invalid %s size', this.branchName)));
   }
   this.size = schema.size | 0;
   this._branchConstructor = this._createBranchConstructor();
@@ -1788,7 +1788,7 @@ FixedType.prototype.random = function () {
 function MapType(schema, opts) {
   Type.call(this);
   if (!schema.values) {
-    throw new Error(f('missing map values: %j', schema));
+    Type.addError(new Error(f('missing map values: %j', schema)));
   }
   this.valuesType = Type.forSchema(schema.values, opts);
   this._branchConstructor = this._createBranchConstructor();
@@ -1878,7 +1878,7 @@ MapType.prototype._write = function (tap, val) {
 };
 
 MapType.prototype._match = function () {
-  throw new Error('maps cannot be compared');
+  Type.addError(new Error('maps cannot be compared'));
 };
 
 MapType.prototype._update = function (rsv, type, opts) {
@@ -1926,7 +1926,7 @@ MapType.prototype._deref = function (schema, opts) {
 function ArrayType(schema, opts) {
   Type.call(this);
   if (!schema.items) {
-    throw new Error(f('missing array items: %j', schema));
+    Type.addError(new Error(f('missing array items: %j', schema)));
   }
   this.itemsType = Type.forSchema(schema.items, opts);
   this._branchConstructor = this._createBranchConstructor();
@@ -2117,10 +2117,10 @@ function RecordType(schema, opts) {
   Type.call(this, schema, opts);
 
   if (!Array.isArray(schema.fields)) {
-    throw new Error(f('non-array record fields: %j', schema.fields));
+    Type.addError(new Error(f('non-array record fields: %j', schema.fields ?? 'no `fields` in schema')));
   }
   if (utils.hasDuplicates(schema.fields, function (f) { return f.name; })) {
-    throw new Error(f('duplicate field name: %j', schema.fields));
+    Type.addError(new Error(f('duplicate field name: %j', schema.fields)));
   }
   this._fieldsByName = {};
   this.fields = Object.freeze(schema.fields.map(function (f) {
@@ -2344,8 +2344,8 @@ RecordType.prototype._createWriter = function () {
 
 RecordType.prototype._update = function (resolver, type, opts) {
   // jshint -W054
-  if (!hasCompatibleName(this, type, !opts.ignoreNamespaces)) {
-    throw new Error(f('no alias found for %s', type.name));
+  if (type.name && !~getAliases(this).indexOf(type.name)) {
+    Type.addError(new Error(f('no alias found for %s', type.name)));
   }
 
   var rFields = this.fields;
@@ -2366,14 +2366,14 @@ RecordType.prototype._update = function (resolver, type, opts) {
       }
     }
     if (matches.length > 1) {
-      throw new Error(
-        f('ambiguous aliasing for %s.%s (%s)', type.name, field.name, matches)
+      Type.addError(new Error())(
+        f('ambiguous aliasing for %s.%s (%s)', type.name, field.name, matches),
       );
     }
     if (!matches.length) {
       if (field.defaultValue() === undefined) {
-        throw new Error(
-          f('no matching field for default-less %s.%s', type.name, field.name)
+        Type.addError(new Error())(
+          f('no matching field for default-less %s.%s', type.name, field.name),
         );
       }
       innerArgs.push('undefined');
@@ -2797,9 +2797,11 @@ AbstractLongType.prototype.compare = utils.abstractFunction;
 
 /** A record field. */
 function Field(schema, opts) {
+  const saveAddError = Type.addError;
+  Type.addError = (err) => { err.fieldName = err.fieldName ? schema.name + '.' + err.fieldName : schema.name; saveAddError(err); };
   var name = schema.name;
   if (typeof name != 'string' || !utils.isValidName(name)) {
-    throw new Error(f('invalid field name: %s', name));
+    Type.addError(new Error(f('invalid field name: %s', name)));
   }
 
   this.name = name;
@@ -2816,12 +2818,15 @@ function Field(schema, opts) {
       case 'ignore':
         return 0;
       default:
-        throw new Error(f('invalid order: %j', order));
+        Type.addError(new Error(f('invalid order: %j', order)));
     }
   })(schema.order === undefined ? 'ascending' : schema.order);
 
   var value = schema['default'];
-  if (value !== undefined) {
+
+  if (value === null && this.type.typeName === 'record') {
+    Type.addError(new Error(f('invalid default for field %s: null is not a %s', this.name, this.type.typeName)));
+  } else if (value !== undefined) {
     // We need to convert defaults back to a valid format (unions are
     // disallowed in default definitions, only the first type of each union is
     // allowed instead).
@@ -2838,7 +2843,7 @@ function Field(schema, opts) {
           type.types[0]
         );
       }
-      throw new Error(msg);
+      Type.addError(new Error(msg));
     }
     // The clone call above will throw an error if the default is invalid.
     if (isPrimitive(type.typeName) && type.typeName !== 'bytes') {
@@ -2848,6 +2853,7 @@ function Field(schema, opts) {
       this.defaultValue = function () { return type._copy(val); };
     }
   }
+  Type.addError = saveAddError;
 
   Object.freeze(this);
 }
@@ -2906,7 +2912,7 @@ function Hash() {
 function readValue(type, tap, resolver, lazy) {
   if (resolver) {
     if (resolver._readerType !== type) {
-      throw new Error('invalid resolver');
+      Type.addError(new Error('invalid resolver'));
     }
     return resolver._read(tap, lazy);
   } else {
@@ -3038,7 +3044,7 @@ function isJsonBuffer(obj) {
  * with a hook since the path is not propagated (for efficiency reasons).
  */
 function throwInvalidError(val, type) {
-  throw new Error(f('invalid %j: %j', type.schema(), val));
+  Type.addError(new Error(f('invalid %j: %j', type.schema(), val)));
 }
 
 function maybeQualify(name, ns) {
@@ -3294,6 +3300,9 @@ function combineObjects(types, opts) {
   return Type.forSchema(schema, opts);
 }
 
+const errorsCollector = [];
+
+Type.addError = (err) => errorsCollector.push(err);
 
 module.exports = {
   Type: Type,
@@ -3313,5 +3322,6 @@ module.exports = {
       types[getClassName(typeName)] = TYPES[typeName];
     }
     return types;
-  })()
+  })(),
+  errorsCollector
 };
